File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Java/ad473362180b0b30cb1e2fa5ba0fdbc0.java
import java.util.*;
import java.io.*;

class Main {
    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }

    static class Pair<T1, T2> {
        T1 first;
        T2 second;

        public Pair(T1 first, T2 second) {
            this.first = first;
            this.second = second;
        }
    }

    static class Triple<T1, T2, T3> {
        T1 first;
        T2 second;
        T3 third;

        public Triple(T1 first, T2 second, T3 third) {
            this.first = first;
            this.second = second;
            this.third = third;
        }
    }

    static class Quadruple<T1, T2, T3, T4> {
        T1 first;
        T2 second;
        T3 third;
        T4 fourth;

        public Quadruple(T1 first, T2 second, T3 third, T4 fourth) {
            this.first = first;
            this.second = second;
            this.third = third;
            this.fourth = fourth;
        }
    }

    static class FastWriter {
        BufferedWriter bw;

        public FastWriter() {
            bw = new BufferedWriter(new OutputStreamWriter(System.out));
        }

        void print(Object object) throws IOException {
            bw.write(object.toString());
        }

        void println(Object object) throws IOException {
            print(object);
            bw.newLine();
        }

        void close() throws IOException {
            bw.close();
        }
    }

    static final int INF = (int) 1e9;
    static final int MAXN = (int) 1e5 + 100;
    static final int MAXNN = (int) 2e4 + 100;
    static final int mod1 = (int) 1e9 + 7;
    static final int mod2 = (int) 1e9 + 21;
    static final long INFLL = (long) 1e18;
    static final int MANX = MAXN;

    static List<Pair<Pair<Integer, Integer>, Pair<Integer, Integer>>> a;
    static int[][] pref;
    static int[] ha, haa_0, haa_1;
    static String ss;
    static List<Integer> p;

    static void ms(int l, int r) {
        if (r <= l)
            return;
        int mid = ((l + r + 1) >> 1) - 1;
        ms(l, mid);
        ms(mid + 1, r);
        a.add(new Pair<>(new Pair<>(l, mid), new Pair<>(mid + 1, r)));
    }

    static void mss(int l, int r) {
        if (r <= l)
            return;
        int mid = ((l + r + 1) >> 1) - 1;
        mss(l, mid);
        mss(mid + 1, r);
        int i = l, j = mid + 1;
        List<Integer> b = new ArrayList<>(r - l + 1);

        int now = 0;
        while (i <= mid && j <= r) {
            if (p.get(i) < p.get(j))
                ss += '0';
            else
                ss += '1';
            b.add(p.get(i));
            i++;
            now++;
        }
        while (i <= mid) {
            b.add(p.get(i));
            i++;
            now++;
        }
        while (j <= r) {
            b.add(p.get(j));
            j++;
            now++;
        }
        for (i = l, j = 0; i <= r; i++, j++)
            p.set(i, b.get(j));
    }

    static void solve() throws IOException {
        FastReader sc = new FastReader();
        FastWriter out = new FastWriter();
        String s = sc.next();
        s = new StringBuilder(s).reverse().toString();
        int n = s.length();
        pref = new int[n + 1][2];
        a = new ArrayList<>();
        p = new ArrayList<>();
        ss = "";
        for (int i = 0; i < n; i++)
            p.add(s.charAt(i) - '0');
        for (int i = 1; i <= n; i++) {
            if (s.charAt(i - 1) == '0') {
                pref[i][0] = pref[i - 1][0] + 1;
                pref[i][1] = pref[i - 1][1];
            } else {
                pref[i][1] = pref[i - 1][1] + 1;
                pref[i][0] = pref[i - 1][0];
            }
        }
        int now = 0;
        int l = 1, r = n;
        while (l <= r) {
            int mid = ((l + r) >> 1);
            int w = check_ans2(mid);
            if (w < n)
                l = mid + 1;
            else
                r = mid - 1;
            now = mid;
        }
        while (true) {
            if (check_ans1(now++))
                break;
        }
        out.close();
    }

    static int check_ans2(int n) {
        a.clear();
        ms(0, n - 1);
        Collections.reverse(a);
        List<Integer> ans = new ArrayList<>(n);
        int now = 0;
        int y = -INF;
        for (int i = 0; i < a.size(); i++) {
            Pair<Pair<Integer, Integer>, Pair<Integer, Integer>> to = a.get(i);
            int sz1 = to.first.second - to.first.first + 1;
            int sz2 = to.second.second - to.second.first + 1;
            int was = now;
            int cur = 1;
            int cnt1 = 0, cnt2 = 0;
            now = n;
            for (int j = now - 1; j >= 0; j--) {
                char u = ss.charAt(j);
                if (u == '0')
                    ans.add(cnt1++, cur++);
                else
                    ans.add(sz1 + cnt2++, cur++);
            }
            while (cnt1 < sz1)
                ans.add(cnt1++, cur++);
            while (cnt2 < sz2)
                ans.add(sz1 + cnt2++, cur++);
            if (i + 1 == a.size())
                y = Math.max(y, n);
        }
        return y;
    }

    static boolean check_ans1(int n) {
        a.clear();
        ms(0, n - 1);
        Collections.reverse(a);
        List<Integer> ans = new ArrayList<>(n);
        int now = 0;
        for (int i = 0; i < a.size(); i++) {
            Pair<Pair<Integer, Integer>, Pair<Integer, Integer>> to = a.get(i);
            int sz1 = to.first.second - to.first.first + 1;
            int sz2 = to.second.second - to.second.first + 1;
            int was = now;
            int cur = 1;
            int cnt1 = 0, cnt2 = 0;
            now = n;
            for (int j = now - 1; j >= 0; j--) {
                char u = ss.charAt(j);
                if (u == '0')
                    ans.add(cnt1++, cur++);
                else
                    ans.add(sz1 + cnt2++, cur++);
            }
            while (cnt1 < sz1)
                ans.add(cnt1++, cur++);
            while (cnt2 < sz2)
                ans.add(sz1 + cnt2++, cur++);
        }
        if (ans.size() != n)
            return false;
        System.out.println(n);
        for (int to : ans)
            System.out.print(to + " ");
        System.out.println();
        return true;
    }

    public static void main(String[] args) throws IOException {
        solve();
    }
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/ad473362180b0b30cb1e2fa5ba0fdbc0.py
from collections import defaultdict
import sys

INFLL = 10**18
MAXN = 10**5 + 100
MAXNN = 2 * 10**4 + 100
INF = 10**9
mod1 = 10**9 + 7
mod2 = 10**9 + 21

a = []
pref = [[0] * 2 for _ in range(MAXN)]
ha = [0] * MAXNN
haa_0 = [0] * MAXNN
haa_1 = [0] * MAXNN

def ms(l, r):
    if r <= l:
        return
    mid = (l + r + 1) // 2 - 1
    ms(l, mid)
    ms(mid + 1, r)
    a.append(((l, mid), (mid + 1, r)))

ss = ""
p = []

def mss(l, r):
    global ss
    if r <= l:
        return
    mid = (l + r + 1) // 2 - 1
    mss(l, mid)
    mss(mid + 1, r)
    i, j = l, mid + 1
    b = [0] * (r - l + 1)

    now = 0
    while i <= mid and j <= r:
        if p[i] < p[j]:
            ss += '0'
            b[now] = p[i]
            now += 1
            i += 1
        else:
            ss += '1'
            b[now] = p[j]
            now += 1
            j += 1
    while i <= mid:
        b[now] = p[i]
        now += 1
        i += 1
    while j <= r:
        b[now] = p[j]
        now += 1
        j += 1
    for i in range(l, r + 1):
        p[i] = b[i - l]

def solve():
    global ss, p
    s = input().strip()
    s = s[::-1]
    for i in range(1, len(s) + 1):
        if s[i - 1] == '0':
            pref[i][0] = pref[i - 1][0] + 1
            pref[i][1] = pref[i - 1][1]
        else:
            pref[i][1] = pref[i - 1][1] + 1
            pref[i][0] = pref[i - 1][0]

    def id3(n):
        global a
        a.clear()
        ms(0, n - 1)
        a.reverse()
        ans = [0] * n
        now = 0

        def does_zero(ff):
            nonlocal now
            cur = 1
            cnt1 = 0
            cnt2 = 0
            to = a[0]
            sz1 = to[0][1] - to[0][0] + 1
            sz2 = to[1][1] - to[1][0] + 1
            cnt1 = 0
            cnt2 = 0
            now = ff
            for i in range(now - 1, -1, -1):
                u = s[i]
                if u == '0':
                    ans[cnt1] = cur
                    cnt1 += 1
                else:
                    ans[sz1 + cnt2] = cur
                    cnt2 += 1
                cur += 1
            while cnt1 < sz1:
                ans[cnt1] = cur
                cnt1 += 1
                cur += 1
            while cnt2 < sz2:
                ans[sz1 + cnt2] = cur
                cnt2 += 1
                cur += 1

        def does(delta, _):
            to = a[_]
            cnt1 = 0
            cnt2 = 0
            sz1 = to[0][1] - to[0][0] + 1
            sz2 = to[1][1] - to[1][0] + 1
            was = now
            fir = [0] * sz1
            sec = [0] * sz2
            cnt1 = 0
            cnt2 = 0
            f = to[0][0]
            now += delta
            for i in range(now - 1, was - 1, -1):
                if s[i] == '0':
                    fir[cnt1] = ans[f]
                    cnt1 += 1
                else:
                    sec[cnt2] = ans[f]
                    cnt2 += 1
                f += 1
            assert cnt1 == sz1 or cnt2 == sz2
            assert cnt1 < sz1 or cnt2 < sz2
            while cnt1 < sz1:
                fir[cnt1] = ans[f]
                cnt1 += 1
                f += 1
            while cnt2 < sz2:
                sec[cnt2] = ans[f]
                cnt2 += 1
                f += 1
            for i in range(to[0][0], to[0][1] + 1):
                ans[i] = fir[i - to[0][0]]
            for i in range(to[1][0], to[1][1] + 1):
                ans[i] = sec[i - to[1][0]]

        dp = [defaultdict(bool) for _ in range(len(a) + 1)]
        dp[0][0] = True
        y = -INF
        for i in range(len(a)):
            if not dp[i]:
                break
            to = a[i]
            sz1 = to[0][1] - to[0][0] + 1
            sz2 = to[1][1] - to[1][0] + 1
            prev_val = -123
            prev_w = -1
            id2 = 0
            id0 = 0
            for w, _ in dp[i].items():
                cnt1 = 0
                cnt2 = 0
                w = w
                if prev_val + 1 == w:
                    cnt1 = id2
                    cnt2 = id0
                    w = prev_w
                    if s[prev_val] == '0':
                        cnt1 -= 1
                    else:
                        cnt2 -= 1
                while w < len(s) and cnt1 < sz1 and cnt2 < sz2:
                    if s[w] == '0':
                        cnt1 += 1
                    else:
                        cnt2 += 1
                    w += 1
                if cnt1 < sz1 and cnt2 < sz2:
                    continue
                prev_val = w
                prev_w = w
                id2 = cnt1
                id0 = cnt2
                if cnt1 == sz1:
                    if s[w] != '0':
                        continue
                else:
                    if s[w] != '1':
                        continue
                if i + 1 == len(a):
                    y = max(y, w)
                dp[i + 1][w] = True
                while w < len(s):
                    if s[w] == '0':
                        if cnt1 + 1 < sz1:
                            cnt1 += 1
                        else:
                            break
                    else:
                        if cnt2 + 1 < sz2:
                            cnt2 += 1
                        else:
                            break
                    w += 1
                    dp[i + 1][w] = True
                    if i + 1 == len(a):
                        y = max(y, w)

        return y

    def id1(n):
        global a
        a.clear()
        ms(0, n - 1)
        a.reverse()
        ans = [0] * n
        now = 0

        def does_zero(ff):
            nonlocal now
            cur = 1
            cnt1 = 0
            cnt2 = 0
            to = a[0]
            sz1 = to[0][1] - to[0][0] + 1
            sz2 = to[1][1] - to[1][0] + 1
            cnt1 = 0
            cnt2 = 0
            now = ff
            for i in range(now - 1, -1, -1):
                u = s[i]
                if u == '0':
                    ans[cnt1] = cur
                    cnt1 += 1
                else:
                    ans[sz1 + cnt2] = cur
                    cnt2 += 1
                cur += 1
            while cnt1 < sz1:
                ans[cnt1] = cur
                cnt1 += 1
                cur += 1
            while cnt2 < sz2:
                ans[sz1 + cnt2] = cur
                cnt2 += 1
                cur += 1

        def does(delta, _):
            to = a[_]
            cnt1 = 0
            cnt2 = 0
            sz1 = to[0][1] - to[0][0] + 1
            sz2 = to[1][1] - to[1][0] + 1
            was = now
            fir = [0] * sz1
            sec = [0] * sz2
            cnt1 = 0
            cnt2 = 0
            f = to[0][0]
            now += delta
            for i in range(now - 1, was - 1, -1):
                if s[i] == '0':
                    fir[cnt1] = ans[f]
                    cnt1 += 1
                else:
                    sec[cnt2] = ans[f]
                    cnt2 += 1
                f += 1
            assert cnt1 == sz1 or cnt2 == sz2
            assert cnt1 < sz1 or cnt2 < sz2
            while cnt1 < sz1:
                fir[cnt1] = ans[f]
                cnt1 += 1
                f += 1
            while cnt2 < sz2:
                sec[cnt2] = ans[f]
                cnt2 += 1
                f += 1
            for i in range(to[0][0], to[0][1] + 1):
                ans[i] = fir[i - to[0][0]]
            for i in range(to[1][0], to[1][1] + 1):
                ans[i] = sec[i - to[1][0]]

        dp = [defaultdict(bool) for _ in range(len(a) + 1)]
        dp[0][0] = True
        for i in range(len(a)):
            if not dp[i]:
                break
            to = a[i]
            for w, _ in dp[i].items():
                cnt1 = 0
                cnt2 = 0
                sz1 = to[0][1] - to[0][0] + 1
                sz2 = to[1][1] - to[1][0] + 1
                w = w
                while w < len(s) and cnt1 < sz1 and cnt2 < sz2:
                    if s[w] == '0':
                        cnt1 += 1
                    else:
                        cnt2 += 1
                    w += 1
                if cnt1 < sz1 and cnt2 < sz2:
                    continue
                if cnt1 == sz1:
                    if s[w] != '0':
                        continue
                else:
                    if s[w] != '1':
                        continue
                dp[i + 1][w] = True
                while w < len(s):
                    if s[w] == '0':
                        if cnt1 + 1 < sz1:
                            cnt1 += 1
                        else:
                            break
                    else:
                        if cnt2 + 1 < sz2:
                            cnt2 += 1
                        else:
                            break
                    w += 1
                    dp[i + 1][w] = True

        if len(dp) > 0 and s not in dp[len(a)]:
            return False

        dp = [defaultdict(bool) for _ in range(len(a) + 1)]
        from_map = [defaultdict(int) for _ in range(len(a) + 1)]
        dp[0][0] = True
        for i in range(len(a)):
            if not dp[i]:
                break
            to = a[i]
            for w, _ in dp[i].items():
                cnt1 = 0
                cnt2 = 0
                sz1 = to[0][1] - to[0][0] + 1
                sz2 = to[1][1] - to[1][0] + 1
                w = w
                while w < len(s) and cnt1 < sz1 and cnt2 < sz2:
                    if s[w] == '0':
                        cnt1 += 1
                    else:
                        cnt2 += 1
                    w += 1
                if cnt1 < sz1 and cnt2 < sz2:
                    continue
                if cnt1 == sz1:
                    if s[w] != '0':
                        continue
                else:
                    if s[w] != '1':
                        continue
                dp[i + 1][w] = True
                from_map[i + 1][w] = w
                while w < len(s):
                    if s[w] == '0':
                        if cnt1 + 1 < sz1:
                            cnt1 += 1
                        else:
                            break
                    else:
                        if cnt2 + 1 < sz2:
                            cnt2 += 1
                        else:
                            break
                    w += 1
                    dp[i + 1][w] = True
                    from_map[i + 1][w] = w

        if len(dp) > 0 and s not in dp[len(a)]:
            return False

        x = []
        now = 0
        ha = len(s)
        for j in range(len(a), 0, -1):
            f = from_map[j][ha]
            x.append(ha - f)
            ha = f
        x.reverse()
        does_zero(x[0])
        assert len(x) == len(a)
        for ii in range(1, len(x)):
            does(x[ii], ii)
        print(n)
        print(" ".join(map(str, ans)))
        return True

    now = 0
    l = 1
    r = 10**5
    while l <= r:
        mid = (l + r) // 2
        w = id3(mid)
        if w < len(s):
            l = mid + 1
        else:
            r = mid - 1
            now = mid
    while True:
        if id1(now):
            return

if __name__ == "__main__":
    input = sys.stdin.read
    solve()
